src/hotspot/share/nmt/memMapPrinter.cpp

```diff
@@ -127,11 +127,21 @@ class CachedNMTInformation : public VirtualMemoryWalker {
   // Given a vma [from, to), find all regions that intersect with this vma and
   // return their collective flags.
   MemFlagBitmap lookup(const void* from, const void* to) const {
+    assert(from <= to, "Sanity");
+    // We optimize for sequential lookups. Since this class is used when a list
+    // of OS mappings is scanned (VirtualQuery, /proc/pid/maps), and these lists
+    // are usually sorted in order of addresses, ascending.
+    static uintx last = 0;
+    if (to <= _ranges[last].from) {
+      // not sequential? restart at 0
+      last = 0;
```

Comment:
If we know that the current region is to the right of the given section, couldn't we just start moving to left, one region at the time, instead of re-starting at 0?

This would be an optimization on top of your optimization, but since we are already touching this might as well consider it?

Reply:
But we don't know how far away the target entry is. We know it precedes the current entry, but not the distance. This is purely theoretical, since in all scenarios we use this class for the input source is sorted. Therefore we cannot predict any likelihood of being "somewhat sorted".

And then, I have the nagging feeling that searching backward would be less cache-friendly. Since there is no pro, and only one possible con, I'd opt for keeping it this way.

Reply:
Yeah, my assumption here was that the sections are somewhat ordered, as per your assumption (which is why your optimization is useful) AND that this miss is closer to the cached index, than 0 (which is unverified assumption on my part).

You would know the best, so if you don't feel like it's worthwhile then I'm OK with that - just wanted to point out this potential additional improvement.